adam/learner/propose_but_verify.py

```diff
@@ -0,0 +1,372 @@
+        if isinstance(perception.frames[0], DevelopmentalPrimitivePerceptionFrame):
+            original_perception_graph = PerceptionGraph.from_frame(
+                perception.frames[0]
+            ).copy_as_digraph()
+        else:
+            raise RuntimeError("Cannot process perception type.")
+        observed_perception_graph = graph_without_learner(original_perception_graph)
+
+        descriptions = []
+        for word, meaning_pattern in self._words_to_hypotheses.items():
+            # Use PerceptionGraphPattern.matcher and matcher.matches() for a complete match
+            matcher = meaning_pattern.matcher(
+                observed_perception_graph, matching_objects=True
+            )
+            if any(
+                matcher.matches(
+                    use_lookahead_pruning=True, graph_logger=self._graph_logger
+                )
+            ):
+                learned_description = TokenSequenceLinguisticDescription(("a", word))
+                descriptions.append((learned_description, 1.0))
+        return immutabledict(descriptions)
+
+    def _log_hypotheses(self, word: str) -> None:
+        assert self._log_word_hypotheses_to
+        # if the user has asked us
+        # to log the progress of the learner's hypotheses about word meanings,
+        # then we use a GraphLogger per-word to write diagram's
+        # of each word's hypotheses into their own sub-directory
+        if word in self._word_to_logger:
+            graph_logger = self._word_to_logger[word]
+        else:
+            log_directory_for_word = self._log_word_hypotheses_to / word
+
+            graph_logger = GraphLogger(
+                log_directory=log_directory_for_word, enable_graph_rendering=True
+            )
+            self._word_to_logger[word] = graph_logger
+
+        def compute_hypothesis_id(h: PerceptionGraphPattern) -> str:
+            # negative hashes cause the dot rendered to crash
+            return str(abs(hash((word, h))))
+
+        if word in self._words_to_hypotheses:
+            logging.info("The word %s has been seen before", word)
+            hypothesised_meaning = self._words_to_hypotheses[word]
+            hypothesis_id = compute_hypothesis_id(hypothesised_meaning)
+            if hypothesis_id not in self._rendered_word_hypothesis_pair_ids:
+                graph_logger.log_graph(
+                    hypothesised_meaning,
+                    logging.INFO,
+                    "Rendering hypothesised " "meaning %s " "for %s",
+                    hypothesis_id,
+                    word,
+                    graph_name=str(hypothesis_id),
+                )
+                self._rendered_word_hypothesis_pair_ids.add(hypothesis_id)
+        else:
+            logging.info(
+                f"The word {word} has never been seen before. There are no current hypotheses for it"
+            )
+
+    # Helper method, from pursuit.py
+    @staticmethod
+    def get_objects_from_perception(
```

Comment:
It would be nice to have this modularized so we can refer to the same function across learners. Let's just keep it this way for now though, as the pursuit_learner.py is also changing right now.